Add dark mode scss

apiraino opened the pull request:
First attempt at avoiding being shock blinded when visiting the blog :smile: 

Note: this CSS applies only to the blog i.e. external links to www.rust-lang.org will take users to `#fff` land for now.

<details><summary>Preview (top)</summary>
<p>

![Screenshot 2024-08-20 at 15-36-43 The Rust Programming Language Blog](https://github.com/user-attachments/assets/baaea6b5-2ea4-44eb-a2de-1c7dfaafbdb8)

</p>
</details> 

<details><summary>Preview (bottom)</summary>
<p>

![Screenshot 2024-06-13 at 20-49-34 The Rust Programming Language Blog](https://github.com/rust-lang/blog.rust-lang.org/assets/6098822/33039a03-dcbe-4477-b6f5-ff85fae79436)

</p>
</details> 

<details><summary>Example blog post</summary>
<p>

![Screenshot 2024-07-29 at 17-17-31](https://github.com/user-attachments/assets/b48067a0-aa9e-40f0-bf46-45144182c3bb)

</p>
</details> 


<details><summary>Mobile view (simulated)</summary>
<p>

![Screenshot 2024-08-20 at 15-38-06 The Rust Programming Language Blog](https://github.com/user-attachments/assets/6277e353-e127-4e08-9824-1883aaec0ec3)



</p>
</details> 

credits to @teohhanhui for providing 80% of this work, thanks!

apiraino commented:
cc @Turbo87 for a review (unsure who is actually in charge for the blog)

thanks

GuillaumeGomez commented on src/styles/app.scss:
```diff
@@ -324,3 +324,55 @@ header h1 {
+  $yellow: #835f00;
+
+  body {
+    background-color: #181a1b;
```
For maintainability reasons, wouldn't it be better to instead define variables and override them in the `@media (prefers-color-scheme: dark)` and have a common CSS ode in any case using these variables? (that's what we do in rustdoc for example)

Turbo87 replied:
another alternative would be using https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/light-dark and potentially have it polyfilled for browsers that don't support it natively yet (that's what we do in crates.io 😉)

GuillaumeGomez replied:
The support is way too recent. Why bother with yet another tool (polyfill) when it can just work? Also, what happens if you want to support more than 2 themes? Anyway, seems interesting but too recent and too limited to be worth it.

apiraino replied:
Using CSS variables is a good suggestion. After thinking about it for a sec, I'd rather do that in a follow-up PR because it's a more invasive change:
- find common colors in both light/dark mode
- find suitable names for them
- refactor the stylesheet

I'm afraid all these changes would block this patch on a review for too long.

GuillaumeGomez replied:
You can start by adding variables for just the cases you want here and we can eventually check for a follow-up later on. What do you think?

apiraino replied:
ok. tried assigning some names. not completely happy but hope it's ok for now

GuillaumeGomez commented on src/styles/app.scss:
```diff
@@ -324,3 +324,58 @@ header h1 {
+  $body-fg: #e8e6e3;
+  $font-color: #d5cbc6;
+
+  body {
```
Hum... You are using SCSS variables when you should use HTML variables. With the current code, you still need to overload all selectors one by one. With what I suggest, you declare variables (kinda like you did) in a `@media` block and that's it.

However, you'd still need to replace the original code with the variables. Let me show an example.

```css
@media (prefers-color-scheme: dark) {
    :root {
        --body-background-color: #181a1b;
        --color: #d5cbc6;
    }
}
```

Then in the original `body`, you replace its content with:

```css
body {
    background-color: var(--body-background-color);
    color: var(--color);
}
```

apiraino commented on src/styles/app.scss:
```diff
@@ -3,14 +3,51 @@
 
-$gray: #2a3439;
-$red: #a72145;
-$green: #0b7261;
-$purple: #2e2459;
```
these 2 colors were not used anywhere, so I removed them

GuillaumeGomez commented on src/styles/app.scss:
```diff
@@ -3,14 +3,51 @@
+  --body-foreground-color: white;
+  --body-color: rgb(34,34,34);
+  --div-brand-a-color: black;
+  --white-color: black;
```
Took me a while to realize it's not "white the color" but "white the element". XD

Please rename it into `--white-elem-color` (or something along the line).

GuillaumeGomez commented on src/styles/app.scss:
```diff
@@ -3,14 +3,51 @@
+  --body-color: rgb(34,34,34);
+  --div-brand-a-color: black;
+  --white-color: black;
+  --white-a: #2a3439;
```
Same here.

GuillaumeGomez commented:
Apart from the two variables with confusing names, it looks all good to me. Great job!

> In a follow-up patch, I'd like to add a button to switch theme. This will unfortunately make the blog not javascript-free anymore, I'm afraid 😞

For me, the important part isn't that it's actually JS-free but rather than it works without JS and that JS only provides limited and secondary improvements. So handling themes with a button doesn't seem that bad to me (and will not require a lot of JS either).

apiraino commented:
ok @GuillaumeGomez I've added a theme selector :slightly_smiling_face: Notes:
- The theme selector is only visible if the user agent has JavaScript enabled (since it's a JavaScript switcher, I thought it was a nice touch hiding it if it's not usable)
- If JavaScript is disabled, we're reading the user agent preferences with no option to change it
- Saves the theme preference in localStorage

GuillaumeGomez commented on src/styles/noscript.scss:
```diff
@@ -0,0 +1,48 @@
+        --nav-links-a: #d5cbc6;
+        --publish-date-author: #d5cbc6;
+        --section-header-h2-color: white;
+        }
```
Broken indent.

GuillaumeGomez commented on static/scripts/theme-switch.js:
```diff
@@ -0,0 +1,31 @@
+    document.documentElement.setAttribute("data-theme", "dark");
+}
+
+// If JavaScript is available enable theme selector
```
```suggestion
// Enable theme selector if javascript is enabled.
```

GuillaumeGomez commented on static/scripts/theme-switch.js:
```diff
@@ -0,0 +1,31 @@
+    document.documentElement.setAttribute("data-theme", "dark");
+}
+
+// If JavaScript is available enable theme selector
+const themeIcon = document.getElementById("theme-icon");
+themeIcon.style.display = 'block';
```
This whole code is useless. You can instead do in `noscript.scss`:

```
#theme-icon {
    display: none !important;
}
```

GuillaumeGomez commented on templates/headers.hbs:
```diff
@@ -29,5 +34,6 @@
- <link type="application/atom+xml" rel="alternate" href="https://blog.rust-lang.org/{{blog.prefix}}feed.xml" title="{{blog.title}}" />
+<!-- atom -->
+<link type="application/atom+xml" rel="alternate" href="https://blog.rust-lang.org/{{blog.prefix}}feed.xml" title="{{blog.title}}" />
+
```
You added a new empty line. Not sure if it's voluntary. ;)

GuillaumeGomez commented on src/styles/app.scss:
```diff
@@ -324,3 +377,37 @@ header h1 {
+  background: var(--theme-popup-bg);
+  position: absolute;
+  list-style: none;
+  margin: 10px 10px 0px 0px;
```
Please remove this. The menu doesn't need margin since it's `position: absolute`.

apiraino replied:
these margins cannot be removed. There is a lot of overlapping styling CSS (I start to think _too much_ styling) and this is the only way to have the theme selection dropdown stay in place.

GuillaumeGomez commented on static/scripts/theme-switch.js:
```diff
@@ -0,0 +1,31 @@
+}
+
+function dropdown () {
+    const themeChoice = document.getElementById("theme-choice");
+    // says it's more "robust" than using getPropertyValue
+    const val = window.getComputedStyle(themeChoice).display
+    if (val === 'none' || val === '') {
+        themeChoice.style.display = 'block';
+    } else {
+        themeChoice.style.display = 'none';
+    }
```
```suggestion
themeChoice.classList.toggle("display");
```

And then in your CSS you add:

```css
#theme-choice.display {
    display: block;
}
```

Do as much as you can in CSS.

GuillaumeGomez replied:
```suggestion
    document.getElementById("theme-choice").classList.toggle("display");
```

And then in your CSS you add:

```css
#theme-choice.display {
    display: block;
}
```

Do as much as you can in CSS.

GuillaumeGomez commented on static/scripts/theme-switch.js:
```diff
@@ -0,0 +1,31 @@
+    // save prefs in the browser
+    localStorage.setItem("blog-rust-lang-org-theme", val);
+    // close the theme dropdown
+    const themeChoice = document.getElementById("theme-choice");
+    themeChoice.style.display = 'none';
```
```suggestion
    document.getElementById("theme-choice").classList.toggle("display");
```

GuillaumeGomez commented on src/styles/app.scss:
```diff
@@ -324,3 +377,37 @@ header h1 {
+}
+
+.theme-icon {
+  display: none;
```
It should not be `display: none` by default, only in the `noscript.css` (I commented about it somewhere else).

GuillaumeGomez commented on src/styles/app.scss:
```diff
@@ -324,3 +377,40 @@ header h1 {
+  position: absolute;
+  list-style: none;
+  margin: 10px 10px 0px 0px;
+  padding: 5px 5px;
```
```suggetsion
    padding: 5px;
```

GuillaumeGomez commented on static/scripts/theme-switch.js:
```diff
@@ -0,0 +1,19 @@
+    // save prefs in the browser
+    localStorage.setItem("blog-rust-lang-org-theme", val);
+    // close the theme dropdown
+    document.getElementById("theme-choice").classList.toggle("display");
```
Sorry, should be:

```suggestion
    document.getElementById("theme-choice").classList.remove("display");
```

GuillaumeGomez commented on templates/nav.hbs:
```diff
@@ -13,5 +13,12 @@
     <li class="tc pv2 ph2 ph4-ns flex-20-s"><a href="https://www.rust-lang.org/tools">Tools</a></li>
     <li class="tc pv2 ph2 ph4-ns flex-20-s"><a href="https://www.rust-lang.org/governance">Governance</a></li>
     <li class="tc pv2 ph2 ph4-ns flex-20-s"><a href="https://www.rust-lang.org/community">Community</a></li>
+    <div class="theme-container">
+      <div id="theme-icon" class="theme-icon" onclick="dropdown();">🖌</div>
+      <ul id="theme-choice" class="theme-choice">
+        <li class="theme-item" onclick="changeThemeTo('light')">Light</li>
+        <li class="theme-item" onclick="changeThemeTo('dark')">Dark</li>
+      </ul>
+    </div>
```
```suggestion
    <div id="theme-icon" class="theme-icon" onclick="dropdown();">🖌
      <ul id="theme-choice" class="theme-choice">
        <li class="theme-item" onclick="changeThemeTo('light')">Light</li>
        <li class="theme-item" onclick="changeThemeTo('dark')">Dark</li>
      </ul>
    </div>
```

GuillaumeGomez commented on static/scripts/theme-switch.js:
```diff
@@ -0,0 +1,61 @@
+}
+
+// close the theme dropdown if clicking somewhere else
+const settingsButton = document.getElementsByClassName('theme-icon')[0];
```
```suggestion
const settingsButton = document.querySelector('.theme-icon');
```

GuillaumeGomez commented on static/scripts/theme-switch.js:
```diff
@@ -0,0 +1,61 @@
+// close the theme dropdown if clicking somewhere else
+const settingsButton = document.getElementsByClassName('theme-icon')[0];
+settingsButton.onblur = handleBlur;
+handleBlur(document.getElementById("theme-choice"));
```
Why do you call `handleBlur` here?

GuillaumeGomez commented on static/scripts/theme-switch.js:
```diff
@@ -0,0 +1,59 @@
+
+// close the theme dropdown if clicking somewhere else
+const settingsButton = document.querySelector('.theme-icon');
+settingsButton.onblur = handleBlur;
```
Normally, `onblur` should be on the button and on the menu from the button. There is one missing here. To test what I mean: open the menu. Click on the theme selector, close it then click somewhere else. Normally, it should close the menu.

GuillaumeGomez commented on static/scripts/theme-switch.js:
```diff
@@ -0,0 +1,64 @@
+}
+
+// close the theme dropdown if clicking somewhere else
+const settingsButton = document.querySelector('.theme-icon');
+settingsButton.onblur = handleBlur;
```
No need to create a new variable here.

```suggestion
document.querySelector('.theme-icon').onblur = handleBlur;
```

apiraino commented:
ok @GuillaumeGomez: as far as I can see - we should be there. I can't think of anything else to add right now.

(I'll update the screenshots in the opening comment to reflect the latest version)

senekor reviewed (approved):
Looks good to me. Some non-blocking thoughts:

- Is it possible to switch to a white version of the Rust logo? The black one is hard to see on the dark background.
- Are there actually people who have dark mode enabled and go around manually reenabling light mode on certain websites? The javascript seems unnecessary to me, but I might be wrong.
- The formatting changes are a bit irritating, but setting up auto-formatting and CI is something for another PR.

Manishearth reviewed (approved):
Can be merged with Remo's comments addressed

GuillaumeGomez commented on src/styles/app.scss:
```diff
@@ -57,6 +57,7 @@ $header-font: 'Alfa Slab One', serif;
   --theme-popup-bg: #141617;
   --theme-hover: #474c51;
   --theme-choice-color: #d5cbc6;
+  --rust-logo-filter: drop-shadow(1px 0 0px #fff) drop-shadow(0 1px 0 #fff) drop-shadow(-1px 0 0 #fff) drop-shadow(0 -1px 0 #fff);
```
You also need to add it for the light theme with this value:

```css
  --rust-logo-filter: initial;
```

GuillaumeGomez commented on src/styles/app.scss:
```diff
@@ -3,14 +3,66 @@
+  --theme-hover: #474c51;
+  --theme-choice-color: #d5cbc6;
+  --rust-logo-filter: drop-shadow(1px 0 0px #fff) drop-shadow(0 1px 0 #fff) drop-shadow(-1px 0 0 #fff) drop-shadow(0 -1px 0 #fff);
+  --rust-logo-filter: initial;
```
Not in here, in the other theme definition.

GuillaumeGomez commented on src/styles/noscript.scss:
```diff
@@ -0,0 +1,53 @@
+  --white-a: #2a3439;
+  --nav-links-a: #2a3439;
+  --publish-date-author: #2a3439;
+  --section-header-h2-color: black;
```
You also need to add the filter in the two themes here.

apiraino commented:
> Looks good to me. Some non-blocking thoughts:
> 
>     * Is it possible to switch to a white version of the Rust logo? The black one is hard to see on the dark background.
> 
>     * Are there actually people who have dark mode enabled and go around manually reenabling light mode on certain websites? The javascript seems unnecessary to me, but I might be wrong.
> 
>     * The formatting changes are a bit irritating, but setting up auto-formatting and CI is something for another PR.

hey @senekor I've fixed the Rust logo and removed the formatting to make it a clean PR.

About the JavaScript to switch theme, I kind of agree with you, I was [hesitant either](https://github.com/rust-lang/blog.rust-lang.org/pull/1343#issuecomment-2256238822) but every Rust site has that (rustdoc, www, not to mention clippy's that without JS is broken) 🤷‍♂️ If you feel strongly about it, happy[^1] to remove it.

[^1]: speaking as someone that usually keeps the NoScript extension on ;-)

Turbo87 commented:
FWIW I was very confused when I visited the blog and it defaulted to dark mode. I double checked by system setting and it was in light mode, but changing it back and forth didn't help. It took a while until I discovered the theme switcher dropdown.

is there a reason why we don't default to the system setting?

GuillaumeGomez commented:
It's supposed to default to light mode as you can see defined in the CSS [here](https://github.com/rust-lang/blog.rust-lang.org/pull/1343/files#diff-1fd86da69dc309c9635372717d250de8288ba6f7a2bfbaa86a01c6a0b1592733R9) and how it's handled in the JS [here](https://github.com/rust-lang/blog.rust-lang.org/pull/1343/files#diff-3093f684cdb04cd798563f52d1f177ef33cc4110b6692e2129b24f0edfe0b4a4R51-R60).

Turbo87 commented:
hmm weird, I guess I must have had an old override in my localstorage somehow. I would still prefer it to follow my system setting though 😅

GuillaumeGomez commented:
We could always add a "system value" option like there is in rustdoc.

apiraino commented:
> hmm weird, I guess I must have had an old override in my localstorage somehow. I would still prefer it to follow my system setting though 😅

The implementation should be:
1. Read user preferences
2. If no user settings, CSS defaults to *system* settings

You should be able to verify with a clean browser profile.

senekor commented:
PR to add system setting: https://github.com/rust-lang/blog.rust-lang.org/pull/1448
